cmd/testbeacon.go

```diff
@@ -451,3 +577,1286 @@ func beaconPeerCountTest(ctx context.Context, _ *testBeaconConfig, target string
+		go singleValidatorSimulation(ctx, duration, target, simulationResCh, params.RequestIntensity, syncCommitteeValidatorsDuties, &wg)
+	}
+
+	log.Info(ctx, "Starting validators performing duties attestation, aggregation, proposal...",
+		z.Any("validators", params.ProposalValidatorsCount),
+	)
+	proposalValidatorsDuties := DutiesPerformed{Attestation: true, Aggregation: true, Proposal: true, SyncCommittee: false}
+	for range params.ProposalValidatorsCount {
+		wg.Add(1)
+		go singleValidatorSimulation(ctx, duration, target, simulationResCh, params.RequestIntensity, proposalValidatorsDuties, &wg)
+	}
+
+	log.Info(ctx, "Starting validators performing duties attestation, aggregation...",
+		z.Any("validators", params.AttestationValidatorsCount),
+	)
+	attesterValidatorsDuties := DutiesPerformed{Attestation: true, Aggregation: true, Proposal: false, SyncCommittee: false}
+	for range params.AttestationValidatorsCount {
+		wg.Add(1)
+		go singleValidatorSimulation(ctx, duration, target, simulationResCh, params.RequestIntensity, attesterValidatorsDuties, &wg)
+	}
+
+	log.Info(ctx, "Waiting for simulation to complete...")
+	// evaluate results
+	wg.Wait()
+	close(simulationGeneralResCh)
+	close(simulationResCh)
+	log.Info(ctx, "Simulation finished, evaluating results...")
+	simulationGeneralRes = <-simulationGeneralResCh
+	for result := range simulationResCh {
+		simulationResAll = append(simulationResAll, result)
+	}
+
+	averageValidatorResult := averageValidatorsResult(simulationResAll)
+
+	finalSimulation := Simulation{
+		GeneralClusterRequests: simulationGeneralRes,
+		ValidatorsRequests: SimulationValidators{
+			Averaged:      averageValidatorResult,
+			AllValidators: simulationResAll,
+		},
+	}
+
+	if !conf.SimulationVerbose {
+		finalSimulation = nonVerboseFinalSimulation(finalSimulation)
+	}
+	simulationResAllJSON, err := json.Marshal(finalSimulation)
+	if err != nil {
+		log.Error(ctx, "Failed to marshal simulation result", err)
+	}
+	err = os.WriteFile(filepath.Join(conf.SimulationFileDir, fmt.Sprintf("%v-validators.json", params.TotalValidatorsCount)), simulationResAllJSON, 0o644) //nolint:gosec
+	if err != nil {
+		log.Error(ctx, "Failed to write file", err)
+	}
+
+	highestRTT := Duration{0}
+	for _, sim := range simulationResAll {
+		if sim.Max.Duration > highestRTT.Duration {
+			highestRTT = sim.Max
+		}
+	}
+	if highestRTT.Duration > thresholdBeaconSimulationPoor {
+		testRes.Verdict = testVerdictPoor
+	} else if highestRTT.Duration > thresholdBeaconSimulationAvg {
+		testRes.Verdict = testVerdictAvg
+	} else {
+		testRes.Verdict = testVerdictGood
+	}
+	testRes.Measurement = highestRTT.String()
+
+	log.Info(ctx, "Validators simulation finished",
+		z.Any("validators_count", params.TotalValidatorsCount),
+		z.Any("target", target),
+	)
+
+	return testRes
+}
+
+func singleClusterSimulation(ctx context.Context, simulationDuration time.Duration, target string, resultCh chan SimulationCluster, wg *sync.WaitGroup) {
+	defer wg.Done()
```

Comment:
Instead of passing `wg *sync.WaitGroup` you could just pass `func()` `wg.Done`

Reply:
Good one, thanks!